lib/puppet/transaction/event_manager.rb

```diff
@@ -144,18 +144,39 @@ def process_callback(resource, callback, events)
       return false
     end
 
+    message = "Triggered '#{callback}' from #{events.length} events"
+
     resource.send(callback)
 
-    if not resource.is_a?(Puppet::Type.type(:whit))
-      resource.notice "Triggered '#{callback}' from #{events.length} events"
-    end
+    resource.notice message if not resource.is_a?(Puppet::Type.type(:whit))
+    event = create_callback_status_event(resource, callback, message) if not resource.is_a?(Puppet::Type.type(:whit))
     return true
   rescue => detail
-    resource.err "Failed to call #{callback}: #{detail}"
+    event = handle_callback_error(resource, transaction, callback, detail)
+    return false
+  ensure
+    if event
+      transaction.resource_status(resource) << event
```

Comment:
This new line is causing this resource to be marked as failed [here](https://github.com/puppetlabs/puppet/blob/4.10.x/lib/puppet/resource/status.rb#L128), which causes the resource [here](https://github.com/puppetlabs/puppet/blob/4.10.x/spec/integration/transaction_spec.rb#L260) to be skipped and this [test](https://github.com/puppetlabs/puppet/blob/4.10.x/spec/integration/transaction_spec.rb#L238) to fail.  Should these resources be failed or not?

Reply:
I think the test might just be wrong. The `exec1` resource is failing (which is expected), and the test is asserting that even though `exec1` has failed, that `exec2` will run. However, `exec2` has declared that it subscribes to `exec1`, which means that `exec1` depends on `exec2`, and `exec2` is getting marked as failed (due to a failed dependency) during [propagate_failure](https://github.com/puppetlabs/puppet/blob/4.10.x/lib/puppet/transaction.rb#L242) when it comes time to evaluate `exec2`. This behavior seems correct to me, and the previous behavior (that allowed the test to pass) seems like the incorrect behavior.

Reply:
it took me a bit to understand why event could be `nil`. I think it'd be cleaner to drop the `ensure` and repeat the logic that adds the event:

```ruby
  if not resource.is_a?(Puppet::Type.type(:whit))
    ...
    event = create_callback_status_event(resource, callback, message)
    transaction.resource_status(resource) << event if event
  end
  return true
rescue => detail
    resource.err "Failed to call #{callback}: #{detail}"
    event = handle_callback_error(resource, transaction, callback, detail)
    transaction.resource_status(resource) << event if event
 end
```